data/transactions/logic/box.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"errors"
+	"fmt"
+)
+
+func opBoxCreate(cx *EvalContext) error {
+	last := len(cx.stack) - 1 // name
+	prev := last - 1          // size
+
+	name := string(cx.stack[last].Bytes)
+	size := cx.stack[prev].Uint
+
+	// This is questionable! We need to think about how boxes can be made during
```

Comment:
@jannotti Sanity checking my understanding:  Does the issue highlighted here occur because the created application ID is _not_ added to `EvalParams.available.boxes` _after_ app creation?

I see in `data/transactions/logic/eval.go:RecordAD` _available_ apps are updated, but not boxes.

Reply:
Yes, I think this can fix this up.  I create the available boxes at EvalParams initialization time, because I want to share boxrefs even from later txns into early ones. So I was ignoring a 0 app id in a box ref unless the app call it was attached to already had an appId (that is, not a creation). But I can add box refs on creation txns during RecordAD once the id is known. Obviously, that means they only apply in later transactions, but that all you could ask for.

To simplify testing, I'll keep a separate table of box refs obtained this way, so that EvalParams.reset() can wipe them.

That improves the story for creation quite a bit. It still means *inner* app creations have no means of referencing boxes, but this is nice progress.